File: framework/src/play-streams/src/main/scala/play/api/libs/streams/ActorFlow.scala
/*
 * Copyright (C) 2009-2016 Lightbend Inc. <https://www.lightbend.com>
 */
package play.api.libs.streams

import akka.actor._
import akka.stream.{ Materializer, OverflowStrategy }
import akka.stream.scaladsl.{ Sink, Keep, Source, Flow }

/**
 * Provides a flow that is handled by an actor.
 *
 * See https://github.com/akka/akka/issues/16985.
 */
object ActorFlow {

  /**
   * Create a flow that is handled by an actor.
   *
   * Messages can be sent downstream by sending them to the actor passed into the props function.  This actor meets
   * the contract of the actor returned by [[akka.stream.scaladsl.Source.actorRef]].
   *
   * The props function should return the props for an actor to handle the flow. This actor will be created using the
   * passed in [[akka.actor.ActorRefFactory]]. Each message received will be sent to the actor - there is no back pressure,
   * if the actor is unable to process the messages, they will queue up in the actors mailbox. The upstream can be
   * cancelled by the actor terminating itself.
   *
   * @param props A function that creates the props for actor to handle the flow.
   * @param bufferSize The maximum number of elements to buffer.
   * @param overflowStrategy The strategy for how to handle a buffer overflow.
   */
  def actorRef[In, Out](props: ActorRef => Props, bufferSize: Int = 16, overflowStrategy: OverflowStrategy = OverflowStrategy.dropNew)(implicit factory: ActorRefFactory, mat: Materializer): Flow[In, Out, _] = {

    val (outActor, publisher) = Source.actorRef[Out](bufferSize, overflowStrategy)
      .toMat(Sink.asPublisher(false))(Keep.both).run()

    Flow.fromSinkAndSource(
      Sink.actorRef(factory.actorOf(Props(new Actor {
        val flowActor = context.watch(context.actorOf(props(outActor), "flowActor"))

        def receive = {
          case Status.Success(_) | Status.Failure(_) => flowActor ! PoisonPill
          case Terminated =>
            println("Child terminated, stopping")
            context.stop(self)
          case other => flowActor ! other
        }

        override def supervisorStrategy = OneForOneStrategy() {
          case _ =>
            println("Stopping actor due to exception")
            SupervisorStrategy.Stop
        }
      })), Status.Success(())),
      Source.fromPublisher(publisher)
    )
  }
}

File: framework/src/play-streams/src/main/java/play/libs/streams/ActorFlow.java
/*
 * Copyright (C) 2009-2018 Lightbend Inc. <https://www.lightbend.com>
 */

package play.libs.streams;

import akka.actor.ActorRef;
import akka.actor.ActorRefFactory;
import akka.actor.Props;
import akka.stream.Materializer;
import akka.stream.OverflowStrategy;
import akka.stream.javadsl.*;
import scala.runtime.AbstractFunction1;

import java.util.function.Function;

/**
 * Provides a flow that is handled by an actor.
 *
 * See https://github.com/akka/akka/issues/16985.
 */
public class ActorFlow {

    /**
     * Create a flow that is handled by an actor.
     *
     * Messages can be sent downstream by sending them to the actor passed into the props function.  This actor meets
     * the contract of the actor returned by {@link akka.stream.javadsl.Source#actorRef}.
     *
     * The props function should return the props for an actor to handle the flow. This actor will be created using the
     * passed in {@link akka.actor.ActorRefFactory}. Each message received will be sent to the actor - there is no back pressure,
     * if the actor is unable to process the messages, they will queue up in the actors mailbox. The upstream can be
     * cancelled by the actor terminating itself.
     *
     * @param <In> the In type parameter for a Flow
     * @param <Out> the Out type parameter for a Flow
     * @param props A function that creates the props for actor to handle the flow.
     * @param bufferSize The maximum number of elements to buffer.
     * @param overflowStrategy The strategy for how to handle a buffer overflow.
     * @param factory The Actor Factory used to create the actor to handle the flow - for example, an ActorSystem.
     * @param mat The materializer to materialize the flow.
     * @return the flow itself.
     */
    public static <In, Out> Flow<In, Out, ?> actorRef(Function<ActorRef, Props> props, int bufferSize, OverflowStrategy overflowStrategy, ActorRefFactory factory, Materializer mat) {

        return play.api.libs.streams.ActorFlow.<In, Out>actorRef(new AbstractFunction1<ActorRef, Props>() {
            @Override
            public Props apply(ActorRef v1) {
                return props.apply(v1);
            }
        }, bufferSize, overflowStrategy, factory, mat).asJava();
    }

    /**
     * Create a flow that is handled by an actor.
     *
     * Messages can be sent downstream by sending them to the actor passed into the props function.  This actor meets
     * the contract of the actor returned by {@link akka.stream.javadsl.Source#actorRef}, defaulting to a buffer size of
     * 16, and failing the stream if the buffer gets full.
     *
     * The props function should return the props for an actor to handle the flow. This actor will be created using the
     * passed in {@link akka.actor.ActorRefFactory}. Each message received will be sent to the actor - there is no back pressure,
     * if the actor is unable to process the messages, they will queue up in the actors mailbox. The upstream can be
     * cancelled by the actor terminating itself.
     *
     * @param <In> the In type parameter for a Flow
     * @param <Out> the Out type parameter for a Flow
     * @param props A function that creates the props for actor to handle the flow.
     * @param factory The Actor Factory used to create the actor to handle the flow - for example, an ActorSystem.
     * @param mat The materializer to materialize the flow.
     * @return the flow itself.
     */
    public static <In, Out> Flow<In, Out, ?> actorRef(Function<ActorRef, Props> props, ActorRefFactory factory, Materializer mat) {

        return play.api.libs.streams.ActorFlow.<In, Out>actorRef(new AbstractFunction1<ActorRef, Props>() {
            @Override
            public Props apply(ActorRef v1) {
                return props.apply(v1);
            }
        }, 16, OverflowStrategy.fail(), factory, mat).asJava();
    }

}
